archunit/src/test/java/com/tngtech/archunit/library/dependencies/SliceRuleTest.java

```diff
@@ -0,0 +1,215 @@
+    }
+
+    private List<String> filterLinesMatching(String text, final String regex) {
+        return FluentIterable.from(Splitter.on(lineSeparator()).split(text))
+                .filter(new Predicate<String>() {
+                    @Override
+                    public boolean apply(String input) {
+                        return input.matches(".*(" + regex + ").*");
+                    }
+                }).toList();
+    }
+
+    private Condition<List<? extends String>> subStringsPerLine(final String... substrings) {
+        return new Condition<List<? extends String>>("substrings per line " + Arrays.asList(substrings)) {
+            @Override
+            public boolean matches(List<? extends String> lines) {
+                if (lines.size() != substrings.length) {
+                    as(description() + ", but number of lines %d does not match number of substrings %d",
+                            lines.size(), substrings.length);
+                    return false;
+                }
+                for (int i = 0; i < lines.size(); i++) {
+                    if (!lines.get(i).contains(substrings[i])) {
+                        as(description() + ", but line %d \"%s\" does not contain substring \"%s\"",
+                                i, lines.get(i), substrings[i]);
+                        return false;
+                    }
+                }
+                return true;
+            }
+        };
+    }
+
+    private String evaluateCompleteGraphCycleFreeWithCycleLimit(int expectedNumberOfCycles) {
+        ArchConfiguration.get().setProperty(MAX_NUMBER_OF_CYCLES_TO_DETECT_PROPERTY_NAME, String.valueOf(expectedNumberOfCycles));
+
+        return getFailureReportForCyclesInRootPackageOf(CompleteSevenNodesGraphRoot.class);
+    }
+
+    private String getFailureReportForCyclesInRootPackageOf(Class<?> packageRoot) {
+        JavaClasses classes = new ClassFileImporter().importPackagesOf(packageRoot);
+        return slices()
+                .matching(packageRoot.getPackage().getName() + ".(*)")
+                .should().beFreeOfCycles().evaluate(classes)
+                .getFailureReport().toString();
+    }
+
+    static int countCyclesInMessage(String text) {
+        return Splitter.on("Cycle detected: ").splitToList(text).size() - 1;
+    }
+
+    /**
+     * To calculate the number of expected cycles check the following (we denote binomial coefficients "n over k" as (n k) ):
+     * For any k-cycle (cycle of k nodes within a graph of n nodes) we can denote
+     * n1 -> n2 -> n3 -> ... -> n[k-1] -> n[k] -> n1
+     * Fix n1 and we have (k-1)! possibilities to order the other k-1 nodes.
+     * Now we have (n k) possibilities to choose k nodes to form a k-cycle and for each choice (k-1)! permutations.
+     * Thus we retrieve the sum of all cycles as
+     * SUM(k=2, n) (n k) * (k-1)! = SUM(k=2, n) n! / ((n-k)! * k)
+     */
+    static int getNumberOfCyclesInCompleteGraph(int numberOfNodes) {
```

Comment:
Is the exact value really needed here?

Reply:
Hey, you just wrote `assertThat(numberOfCycles).isEqualTo(1442525)` :rofl:
Yes, I think I would want to assert it once blackbox against the rule text, just to assert that really all cycles make it into the message. You can argue about the performance test :wink:

Reply:
That makes to me (see also below)